### ISSUE TITLE
Unable to debug JS if debug is on

### ISSUE THREAD

**Issue by alex7r** [joomla/joomla-cms#16381]
### Steps to reproduce the issue
Set 'debug' option to 'on'.
Try to set breakpoints in JS file and reload the page to trigger breakpoint.

### Expected result
Browser stops at the breakpoint.

### Actual result
Browser never see breakpoint.

### Additional comments
Issue is caused by:
Debug option enables "nocache" versioning. So no matter if you set breakpoint at resource - there would be new resource on next page load.
```
/media/editors/tinymce/js/tinymce.js?2516a3ab5432c9a021777999ae2f7f36
/media/editors/tinymce/js/tinymce.js?679fa0a9f9cffcfdf7528cce440a8587
/media/editors/tinymce/js/tinymce.js?e124197480d0cb6139dd3a0196ec953b
```
And without debug everything is fine except that there is only minified files. 
```
/media/editors/tinymce/js/tinymce.min.js?63cc71938d046f0cc7b2692e664316fe
/media/editors/tinymce/js/tinymce.min.js?63cc71938d046f0cc7b2692e664316fe
```

**Comment by zero-24** [comment]
can't you insert the debug breake point inside of the js file?

**Comment by alex7r** [comment]
you mean edit file in file system? you can, but why would you if browsers have extended functionality for this? shouldn't​ we support worldwide used development tools and tech?

**Comment by yvesh** [comment]
Ping @dgt41

**Comment by tonypartridge** [comment]
This is an interesting point, since I believe the versioning was added to allow a full refresh of JS when enabling debug since many users forget to clear then cache we make sure they get a new file.

I suppose we could add an option to the debug plugin that adds the version to JS file or not?

Thoughts @dgt41 ?

**Comment by alex7r** [comment]
no, both files​ are shipped. 
It's what Tony described above.
but I believe there is a way to make this work - version should be set based on modified timestamp.

**Comment by tonypartridge** [comment]
By that do you mean the time stamp that the debug mode was enabled? I still think there is a usage case for it loading a new file every time when debug mode is enabled.

**Comment by alex7r** [comment]
no, timestamp from filesystem, file modified time

**Comment by alex7r** [comment]
I think that's really the way to handle versioning and cache.

**Comment by tonypartridge** [comment]
Ah ok, yep that to me would be a good idea but only when debug mode is enabled.

**Comment by brianteeman** [comment]
sorry i thought you were referring to the js from tinymce not the joomla js

**Comment by alex7r** [comment]
Tony, do you think that checking file modified date with filesystem will cause performance impact?

**Comment by tonypartridge** [comment]
It will have an effect given it's an extra process but your debugging so expect it a bit slower anyhow. Should be worth a test nontheless.

**Comment by andrepereiradasilva** [comment]
i'm not in favour of I/O file system calls to check file timestamps ...
a lot of I/O fs calls can be a performance problem in servers using the `open_basedir` php ini config option - normally shared servers - since for security reasons when you enable this option php does not cache filesystem paths - reference: https://bugs.php.net/bug.php?id=52312

also the https server and browser should already be doing that using HTTP protocol Headers (ex: "Last-modified", "ETag") 

but you can also disable appending the media version to scripts/stylesheets when in debug mode (using JDEBUG const)

if anyone is interested the code that adds the media version is in:
- Gets media version https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/document/renderer/html/head.php#L70
- Adds media version in stylesheets https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/document/renderer/html/head.php#L144
- Adds media version in scripts https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/document/renderer/html/head.php#L255

**Comment by brianteeman** [comment]
+10

**Comment by tonypartridge** [comment]
Agreed, I just checked JFile out, I sheepishly assumed it had a an info attribute containing basic info like JImage.

Anyhow, then my solution still stands. Not many people are debugging JS IMHO within the browser. So for the ones who are they know how to clear a cache.

We just need to allow adding and removing of the unique version numbering, within the debug plugin I think should suffice.

**Comment by ggppdk** [comment]
@andrepereiradasilva 

> i'm not in favour of I/O file system calls to check file timestamps ...
> a lot of I/O fs calls can be a performance problem in servers using the open_basedir php ini config option - normally shared servers - since for security reasons when you enable this option php does not cache filesystem paths - reference: https://bugs.php.net/bug.php?id=52312

I had same thoughts, so in my extension in order to avoid it
i use a timestamp from a single file to version all files,

It is a file that is modified on every extension upgrade ...

This way i avoids making a few dozens of time stamp checks , 1 per file,
and instead always doing only 1

This requires to hard refresh pages during developing time (or an option to generate new version),

but it is safe for refreshing caching of JS / CSS file for users upgrading my extension, and also works with setting breakpoint in the browser

**Comment by tonypartridge** [comment]
But that's pointless in this usage case, because my argument is fine with a config option. Since we could use the modified time from when the config was saved for instance and debug was enabled.

But I don't think that's necessary and we are just overly complicating things.

**Comment by ggppdk** [comment]
> But that's pointless in this usage case, because my argument is fine with a config option.

Not pointless since my suggestion works, if there are alternatives, it does not mean that something is pointless

> Since we could use the modified time from when the config was saved for instance and debug was enabled.

This will not work in all cases

**Comment by ggppdk** [comment]
> This will not work in all cases

i meant if you upgrade and configuration was saved / set before upgrading, plus that solution would a little more work to implement, and some argueably unneeded checks will always be running on save of global configuration

but i would not be against it, whatever

**Comment by andrepereiradasilva** [comment]
@ggppdk 
> I had same thoughts, so in my extension in order to avoid it
> i use a timestamp from a single file to version all files,

just additional info: the joomla media version does not do any timestamp or fs I/O at all.
Is changed, for instance, on joomla update, or whenever `JFactory::getApplication()->flushAssets();` (otr the code of this method) is called
see
- https://github.com/joomla/joomla-cms/search?q=flushAssets
- https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/application/web.php#L967-L971
- https://github.com/joomla/joomla-cms/blob/staging/libraries/cms/version/version.php#L244-L325

and also could be changed by any 3º party extension.

example on joomla patch tester: https://github.com/joomla-extensions/patchtester/pull/178/files (in this case change the joomla media version whenever you apply a patch to avoid js/css browser cache issues)

**Comment by alex7r** [comment]
Tony, If I understand you correctly, than checking config file timestamp is not a solution as it leads to caching files from time of enabling debug. And this extra effort for same result as just strip version away.
And if file checks are bad for performance than "no version" is the best solution here.

**Comment by ggppdk** [comment]
@andrepereiradasilva 

you are right, just since it is updated by Joomla upgrade it is not suitable for extensions

but then as you say...

> and also could be changed by any 3º party extension, ... example on joomla patch tester: 

i did not consider doing this !
... probably it is not a bad choice to do for my extension too

Just invalidating client caching of many site JS / CSS files, not related to my extension does not sound necessary

Now about this topic, a simple solution is to update the versioning URL hash number more often
- not only on Joomla upgrade (done already)
- but also on global configuration saving regardless if debug setting is ON, 
this will not only solve this issue, but also help speeding up working with debug on,
just if you modify JS files you will need to do hard refresh every time

**Comment by andrepereiradasilva** [comment]
maybe [onAfterDispatch](https://github.com/joomla/joomla-cms/blob/staging/plugins/system/debug/debug.php#L192) debug system plugin

if ((boolean) joomla global debug === true && (boolean) new param "refresh media version" === true)
then
`JFactory::getApplication()->flushAssets();`
end if

:)

**Comment by ggppdk** [comment]
Global configuration saving does not the event 'onExtensionBeforeSave'
`public function onExtensionBeforeSave($context, $table, $isNew)
`

Maybe better add triggering of the above (with an appropriate "context") ?
it will be useful in other cases too

**Comment by alex7r** [comment]
I lost track of it... so you propose to update version when debug is on? and the issue here that it's happening and preventing browser from setting breakpoint...

**Comment by stevlam** [comment]
@alex7r i Lost track of the issue itself sorry
Will check better when i have more time

**Comment by andrepereiradasilva** [comment]
@alex7r  will check when i have more time

**Comment by andrepereiradasilva** [comment]
we could also disable the media version if debug is enabled, but imo it should be enabled so in debug mode we always use a new file.

**Comment by andrepereiradasilva** [comment]
ok, so agree with @tonypartridge is IMO a good solution for this particular case

> We just need to allow adding and removing of the unique version numbering, within the debug plugin I think should suffice.

**Comment by andrepereiradasilva** [comment]
see https://github.com/joomla/joomla-cms/pull/16406

**Comment by franz-wohlkoenig** [comment]
closed as having PR #16406<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16381">issues.joomla.org/tracker/joomla-cms/16381</a>.</sub>
